### DISCUSSION TITLE
v9.0.0 released

### DISCUSSION THREAD

**Discussion by gpbl** [gpbl/react-day-picker#2280] [Announcements]
DayPicker v9 is a major release including significant updates related to accessibility, customization and localization. 

- See the updated docs at https://daypicker.dev and the [upgrading Guide](https://daypicker.dev/upgrading) for more details.

### Install the Latest Version

```bash
npm install react-day-picker@latest
```

## What’s New

- Moved `date-fns` from peer dependencies to dependencies.
- Added support for [UTC dates](https://daypicker.dev/docs/localization#utc-dates) and [Jalali Calendar](https://daypicker.dev/docs/localization#jalali-calendar).
- [Enhanced accessibility](https://daypicker.dev/docs/accessibility) to better comply with [WCAG 2.1](https://www.w3.org/TR/WCAG21/) recommendations.
- [Simplified styles](https://daypicker.dev/docs/styling) and new CSS variables for easier customization.
- Improved selection logic for [range mode](https://daypicker.dev/docs/selection-modes.mdx).
- New `dropdown-years` and `dropdown-months` caption layouts.
- New `hideWeekdayRow` and `hideNavigation` props.
- Updated for a complete [custom components](https://daypicker.dev/guides/custom-components) support.
- Improved typings and props for better compatibility in TypeScript `strict` mode.

## Breaking Changes

While we tried to keep the API as stable as possible, some breaking changes were necessary to improve the library:

- The updated ARIA labels could require new translations or updated unit tests selectors.
- Custom CSS styles will likely break, due to the updated CSS classes and simplified styles.
- Custom Components have new API and may break.
- Some typings have been renamed or deprecated.
- The `useInput` hook has been removed. See [Input fields](https://daypicker.dev/guides/input-fields) guide for more details.
- `onWeekNumberClick` has been removed. Use a custom component to handle week number clicks.
- The updated build system to ESM and CommonJS could break some custom bundler.

### Upgrading Guide

We prepared a [Upgrading guide](https://daypicker.dev/upgrading) for help upgrading your app to v9. We welcome [feedback](https://github.com/gpbl/react-day-picker/discussions) about the upgrade process, to ensure it's smooth for everyone.

## Compatibility

DayPicker v9 is compatible with React 16.8+.

## Get Support and Report Issues

Get support, report issues, and provide feedback on the [Discussion forums](https://github.com/gpbl/react-day-picker/discussions). Thanks.

## New Contributors

* @grzegorzpokorski made their first contribution in https://github.com/gpbl/react-day-picker/pull/2124
* @zaaakher made their first contribution in https://github.com/gpbl/react-day-picker/pull/2137
* @ArthurGoupil made their first contribution in https://github.com/gpbl/react-day-picker/pull/2230
* @johnnysedh3lllo made their first contribution in https://github.com/gpbl/react-day-picker/pull/2252
* @xardit made their first contribution in https://github.com/gpbl/react-day-picker/pull/2254
* @dave-meyer made their first contribution in https://github.com/gpbl/react-day-picker/pull/2231
* @benasher44 made their first contribution in https://github.com/gpbl/react-day-picker/pull/2249
* @RyanCavanaugh made their first contribution in https://github.com/gpbl/react-day-picker/pull/2275

**Full Changelog**: https://github.com/gpbl/react-day-picker/compare/v8.10.1...v9.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/gpbl/react-day-picker/releases/tag/v9.0.0'>v9.0.0</a>.</em>

**Comment by AhmedBaset** [comment]
Great job @gpbl 

Too many `react-day-picker` consumers use it via the [Shadcn `<Calendar />`](https://ui.shadcn.com/docs/components/calendar). Would you mind opening a PR there migrating from v8 to v9? so v8 users can copy the updated component.

I tried migration from 8 to 9rc-5 a couple days ago and had difficult experience with no success at the end. You are the best one know what has changed.

**Reply by TheMikeyRoss** [reply]
Also a hijri feature would be nice

**Reply by gpbl** [reply]
> I tried migration from 8 to 9rc-5 a couple days ago and had difficult experience with no success at the end. You are the best one know what has changed.

Hi @AhmedBaset, thanks for trying it out - do you had difficulties upgrading react-day-picker used via  `shadcn/ui` or with the calendar itself?

> Too many `react-day-picker` consumers use it via the [Shadcn `<Calendar />`](https://ui.shadcn.com/docs/components/calendar). Would you mind opening a PR there migrating from v8 to v9? so v8 users can copy the updated component.

Their [calendar.tsx](https://github.com/shadcn-ui/ui/blob/main/apps/www/registry/default/ui/calendar.tsx) is using components and class names that have been removed. We may reintroduce (and deprecate) them to make it compatible with their code, but I cannot guarantee the style would look the same.

**Reply by TheMikeyRoss** [reply]
Im actually still stuck trying to migrate to v9 with the same look as the shadcn/ui look.

What replaced `table` classname? 

Could you kindly remake the shadcn calendar but with the new styling names and classnames and components of v9? it will help a lot of people, and it will also help update the current one in shadcn to v9

**Reply by AhmedBaset** [reply]
The difficulty was in migrating the calendar.tsx file similar to @TheMikeyRoss experience, what's the replacement of nav_button, and many more classNames. I often look at the source code to see what's been changed but I got stuck in #2161 500+ changed file and the upgrade guide doesn't mention all of these.

**Comment by gpbl** [comment]
> Also a hijri feature would be nice

@TheMikeyRoss We need to rely on third-party packages for supporting other calendars, as we do for date-fns-jalali.

In the JS ecosystem see only https://www.npmjs.com/package/hijri-date and its [HijriDate.js](https://github.com/abdennour/hijri-date/blob/master/src/HijriDate.js). 

I have no experience with this calendar - is an updated Date constructor as in HijriDate enough? If yes, then this should work:

```tsx
import HijriDate from 'hijri-date';

<DayPicker dateLib={{ Date: HijriDate }} />
```

**Comment by efgomes** [comment]
![image](https://github.com/user-attachments/assets/11946669-e694-4e5e-8778-5083c1ce97e4)


what is the new class to replace the old nav_button?

**Reply by gpbl** [reply]
> what is the new class to replace the old nav_button?


Hi @efgomes, thanks for your feedback. 

We just released a patch that should make the upgrade easier. 

Try to install `9.0.1` and see if the updated docs can help:

https://daypicker.dev/upgrading#3-update-the-class-names

```
npm install react-day-picker@latest
```

**Comment by TheMikeyRoss** [comment]
@gpbl I think you'll be bombarded with devs having issues specifically with the incompatible classnames with shadcn-ui 😅

I think the best course of action is to make a code snippet of the new way to replicate shadcn-ui/calendar and pin it in the discussion board.

I wish I could help with that but I'm also stuck with nav buttons, table, rows, cells, and pretty much 60% of the deprecated classes.

**Reply by AhmedBaset** [reply]
As it went with cmdk@1 and `<Command />`

**Reply by gpbl** [reply]
> I wish I could help with that but I'm also stuck with nav buttons, table, rows, cells, and pretty much 60% of the deprecated classes.

Hi @TheMikeyRoss, @AhmedBaset - thanks for your feedback. I just released a patch that should make the upgrade easier. 

Try to install `v9.0.1` and see if the updated docs can help:

https://daypicker.dev/upgrading#3-update-the-class-names

```
npm install react-day-picker@latest
```

> @gpbl I think you'll be bombarded with devs having issues specifically with the incompatible classnames with shadcn-ui 😅
> 
> I think the best course of action is to make a code snippet of the new way to replicate shadcn-ui/calendar and pin it in the discussion board.

I'm not familiar with `shadcn-ui`: do they have a way to handle mayor upgrades to their dependencies? Hopefully somebody will push an update to their repo (I will probably busy handling the new feedbacks here) 🙏🏽

**Reply by zaaakher** [reply]
@gpbl `shadcn-ui` isn't really a package, and it has no dependencies. It's simply a collection of copyable components (or automated copy via shadcn cli). 
So someone can go to the [calendar component](https://ui.shadcn.com/docs/components/calendar) and follow the manual instructions to create the Calendar component

![image](https://github.com/user-attachments/assets/203029ce-368a-49f6-ac34-640567de7d7f)

And that install command will use the latest version of `react-day-picker`. And if you scroll down you'll see the code to copy, and when someone copies it they'll see ts warnings and issues due to the v9 breaking changes of classes.

So all we have to do to mitigate people having issues is to make a PR to shadcn to update the code snippet to work for v9. (while keeping the looks and feels identical) in [this file](https://github.com/shadcn-ui/ui/blob/main/apps/www/registry/default/ui/calendar.tsx)

**Reply by AhmedBaset** [reply]
I opened a WIP PR https://github.com/shadcn-ui/ui/pull/4371

**Reply by gpbl** [reply]
Thanks here for the insights about shadcn-ui. It seems the missing version in their install instructions is giving trouble to many. 

Isn't easier fix their doc to include the version of the package?

```diff
- npm install react-day-picker
+ npm install react-day-picker@8
```

**Reply by TheMikeyRoss** [reply]
I'd consider that a temporary bandaid. someone running a a tool like [node-check-update](https://github.com/raineorshine/npm-check-updates) might bumb rdp to v9 and they're back to square one. 

The unfortuante thing about shadcn is that the maintainer shows up once a week. As you can see there's almost 600 PRs. So even if we make a PR for the `npm install react-day-picker@8` it won't be merged for some time. So we might as well just migrate the whole component to v9

**Reply by gpbl** [reply]
> I'd consider that a temporary bandaid. someone running a a tool like [node-check-update](https://github.com/raineorshine/npm-check-updates) might bumb rdp to v9 and they're back to square one.

Breaking change between mayor versions are expected.

`shadcn-ui` uses third party packages in their examples, but they miss saying which version is their code compatible to. IMHO it is just that.

**Comment by TheMikeyRoss** [comment]
It seems the `day_button` classes overwrite `selected`

@AhmedBaset Are you facing something similar? where the selected day changes to light gray when you hover on it? (from shadcn)

**Reply by zaaakher** [reply]
I'm facing that exact issue. See https://github.com/sikka-software/Hawa/pull/64

**Reply by gpbl** [reply]
v8 was rendering the elements for a the grid cell roughly as:

```html
<td>
  <span class="selected today">
    <button>{date}</button>
  </span>
</td>
```

...making difficult to style the `td` element when a day is selected.

From v9 it renders something as:

```html
<td class="selected today"> // ✅ now this element can be styled too
  <button>{Date}</button>
</td>
```

This may cause your problems @TheMikeyRoss @zaaakher. Give a look to the HTML source of your code to understand better the new HTML hierarchy.

**Reply by zaaakher** [reply]
@TheMikeyRoss I solved it by simply replacing the `DayButton` component with a custom one. And you can use `props.modifiers.selected` to conditionally render different styles when a day is selected.

https://github.com/sikka-software/Hawa/blob/7677a39689cb21e6da1d6176c9f7032cee380737/packages/components/elements/calendar/Calendar.tsx#L63-L73

**Reply by gpbl** [reply]
> @TheMikeyRoss I solved it by simply replacing the `DayButton` component with a custom one. And you can use `props.modifiers.selected` to conditionally render different styles when a day is selected.

Love this solution, thanks for sharing. 

Is this problem related to this line in style.css? 

https://github.com/gpbl/react-day-picker/blob/330ae8e4b15eab00fcf0a938d0b0815beeab77f0/src/style.css#L251-L253

I understand targeting a child just via classnames is creating the issue here.

**Reply by zaaakher** [reply]
> Is this problem related to this line in style.css?

It could be. But I don't understand why it would be.

I'd summarize the issue as "The hover styles of `day_button` overwrite the styles of `selected`"

**Reply by gpbl** [reply]
🤔 Then the issue may be the "button reset": 

https://github.com/gpbl/react-day-picker/blob/330ae8e4b15eab00fcf0a938d0b0815beeab77f0/src/style.css#L75-L84

Let see if other meets similar issues. I will keep an eye open on this, thanks again for reporting.

**Reply by zaaakher** [reply]
> Then the issue may be the "button reset":

Ah that looks more likely.

> Let see if other meets similar issues. I will keep an eye open on this, thanks again for reporting.

Great, thanks for all your work 👍

**Reply by berkerdemirer** [reply]
_I understand targeting a child just via classnames is creating the issue here._

I am currently, having this issue. If you would like to have range select as same as in provided in the docs using tailwind and classNames approach, it won't work. Maybe I am missing something here but how can you target `.range_start .day_button` with only given `range_start` classname key? Due to that, when you give background color to `range_start` it will actually override whatever you add for `select`.
